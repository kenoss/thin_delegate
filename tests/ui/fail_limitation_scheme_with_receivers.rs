#[thin_delegate::register]
pub trait Hello {
    fn hello_ref(&self, prefix: &str) -> String;
    fn hello_ref_mut(&mut self, prefix: &str) -> String;
    fn hello_consume(self, prefix: &str) -> String;
}

impl Hello for usize {
    fn hello_ref(&self, prefix: &str) -> String {
        format!("{prefix}, {self}")
    }

    fn hello_ref_mut(&mut self, prefix: &str) -> String {
        format!("{prefix}, {self}")
    }

    fn hello_consume(self, prefix: &str) -> String {
        format!("{prefix}, {self}")
    }
}

#[thin_delegate::register]
struct Hoge(String);

impl Hoge {
    fn key(&self) -> usize {
        self.0.len()
    }
}

// `#[thin_delegate::fill_delegate]` can't handle consuming receiver with `scheme` argument.
// You need to implement such functions manually.
//
// See also pass_scheme_with_receivers.rs.
#[thin_delegate::fill_delegate(scheme = |f| f(&mut self.key()))]
impl Hello for Hoge {}

fn main() {}
