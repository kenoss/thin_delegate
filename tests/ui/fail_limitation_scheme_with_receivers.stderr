error[E0277]: the trait bound `&mut usize: Hello` is not satisfied
  --> tests/ui/fail_limitation_scheme_with_receivers.rs:35:47
   |
5  |     fn hello_consume(self, prefix: &str) -> String;
   |        ------------- required by a bound introduced by this call
...
35 | #[thin_delegate::fill_delegate(scheme = |f| f(&mut self.key()))]
   |                                               ^^^^^^^^^^^^^^^ the trait `Hello` is not implemented for `&mut usize`
   |
   = note: this error originates in the macro `__thin_delegate__trampoline2` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider removing the leading `&`-reference
   |
35 - #[thin_delegate::fill_delegate(scheme = |f| f(&mut self.key()))]
35 + #[thin_delegate::fill_delegate(scheme = |f| f(self.key()))]
   |
